object/virtual_device_list.go

```diff
@@ -361,6 +373,77 @@ func (l VirtualDeviceList) newNVMEBusNumber() int32 {
 	return -1
 }
 
+// FindSATAController will find the named SATA or AHCI controller if given, otherwise will pick an available controller.
+// An error is returned if the named controller is not found or not a SATA or AHCI controller. Or, if name is not
+// given and no available controller can be found.
+func (l VirtualDeviceList) FindSATAController(name string) (types.BaseVirtualController, error) {
+	if name != "" {
+		d := l.Find(name)
+		if d == nil {
+			return nil, fmt.Errorf("device '%s' not found", name)
+		}
+		switch c := d.(type) {
+		case *types.VirtualSATAController:
+			return c, nil
+		case *types.VirtualAHCIController:
+			return &c.VirtualSATAController, nil
+		default:
+			return nil, fmt.Errorf("%s is not a SATA or AHCI controller", name)
+		}
+	}
+
+	c := l.PickController((*types.VirtualSATAController)(nil))
+	if c == nil {
+		c = l.PickController((*types.VirtualAHCIController)(nil))
+	}
+	if c == nil {
+		return nil, errors.New("no available SATA or AHCI controller")
+	}
+
+	switch c := c.(type) {
+	case *types.VirtualSATAController:
+		return c, nil
+	case *types.VirtualAHCIController:
+		return &c.VirtualSATAController, nil
```

Comment:
This should return `c`, not a pointer to the embedded field, right?

Reply:
Resolved in 2e77836.